### PR TITLE
Services Developer Guide

### LINKED ISSUE #610: Publish an infrastructure stack developer guide
Make it easy for new contributors to [a] extend existing infra stacks to add support for additional managed cloud services (Kafka, Spark, Hadoop), and [b] create new infra stacks for additional cloud providers (Alibaba Cloud, Tencent, Digital Ocean).

Part of the [0.3 release](https://github.com/crossplaneio/crossplane/blob/master/ROADMAP.md#v03---enable-partners-to-build-infra-stacks)

### Why? What problem are we solving?
A [Stack](https://github.com/crossplaneio/crossplane/blob/60333b/design/design-doc-stacks.md) is a set of CRDs and controllers that extend Crossplane. An infrastructure stack adds support for infrastructure (e.g. databases, VMs, etc), as opposed to adding support for an application (e.g. Wordpress, GitLab). Crossplane's existing managed resource functionality could be broken up into GCP, Azure, and AWS infrastructure stacks.

We hope that breaking Crossplane's infrastructure managed functionality up into stacks will **make it easier for the community to write and maintain their own infrastructure stacks**. Crossplane will likely provide some guide rails in the form of libraries and interfaces with which to write controllers, but we believe this would be best supplemented by great documentation.

### What does it look like when we're done?

* We have published an end-to-end guide to writing an infrastructure stack to https://crossplane.io/.
* Our guide assumes familiarity with Kubernetes and Go, but not kubebuilder. It prefers linking to kubebuilder documentation over re-documenting it.
* Our guide covers at least:
  * Why write an infrastructure stack?
  * Crossplane concepts and terminology including managed resource, resource class, default resource classes, resource claim, external resource, etc.
  * Any patterns any opinions Crossplane recommends on top of the existing kubebuilder documentation, for example fields required by all Crossplane CRDs, and recommended `Reconciler` scaffolds such as the resource claim, managed resource, and default resource class reconciler libraries.
  * How to build and package infrastructure stacks.

### How could this be demonstrated?

1. Crossplane user reads the blog post describing this new functionality
1. Click through to the updated user guide
1. Try out the updated example(s)

For example:
**Scenario 1**: add new infra controller to existing infra stack in the crossplaneio org (TBD)
e.g.
* raise issue & get community feedback
* create CRD
* create controller
* submit PR, what we look for in new controller

**Scenario 2**: create new infra stack for unsupported cloud (TBD)
e.g.
* copy the GCP infra stack to a new repo
* adapt cloud specific areas
* publish to registry
* recommended testing:
  * validate portable wordpress app stack deployment #572 
  * etc.

### Related
* Crossplane Glossary & Concepts (see #625 features as well, should be the same)
* Enable users to extend existing infra stacks by adding a new controllers & CRDs
  * TBD, @negz 
  * controller patterns - docs, how reconcilers are designed
  * stack experience, registry, publishing, contents of infra stack, installed to control plane, upgrades
* Enable partners to add their own infra stack & CRDs
* Advanced controller concept, patterns, tech deep dive: incorporate [Illya's reconciler patterns](https://github.com/crossplaneio/crossplane/blob/master/design/design-doc-reconciler-patterns.md)
* Short blog post to blog.crossplane.io about this new functionality! #TBD

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

Fixes #610

### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the RBAC permissions in [`clusterrole.yaml`] to include any new types.

[documentation]: https://github.com/crossplaneio/crossplane/tree/master/docs
[examples]: https://github.com/crossplaneio/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplaneio/crossplane/tree/master/PendingReleaseNotes.md
[`clusterrole.yaml`]: https://github.com/crossplaneio/crossplane/blob/master/cluster/charts/crossplane/templates/clusterrole.yaml

### CODE REVIEW

File: docs/adding-a-managed-service.md [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+}
+
+// +kubebuilder:object:root=true
+
+// A FavouriteDBInstanceClass is a non-portable resource class. It defines the
+// desired spec of resource claims that use it to dynamically provision a
+// managed resource.
+type FavouriteDBInstanceClass struct {
+	metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
```
Comment (jbw976) [inline docs/adding-a-managed-service.md]: indentation on this line is less than the other fields in this struct

File: docs/adding-a-managed-service.md [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+// the resource.Claim interface in the interest of brevity. These methods all
+// set or get a field of one of the embedded structs, similar to SetBindingPhase
+// above.
+```
+
+### Portable Resource Class Kinds
+
+Portable resource classes are typically defined alongside the resource claim
+they align with, similar to how non-portable resource classes are defined
+alongsie the managed resources they align with.
```
Comment (jbw976) [inline docs/adding-a-managed-service.md]: typo: alongside

File: docs/adding-a-managed-service.md [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+
+Portable resource classes must:
+
+* Satisfy crossplane-runtime's [`resource.PortableClass`] interface.
+* Directly embed a [`PortableClass`] struct.
+* Have no `Spec` field.
+* Have no `Status` Field`.
```
Comment (jbw976) [inline docs/adding-a-managed-service.md]: hanging/unmatched backtick at the end of this line

File: docs/adding-a-managed-service.md [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+
+## Adding controllers
+
+Write some code.
+
+## In Review
+
+We wrote a lot of code.
+
+[managed resource]: glossary.md#managed-resource
```
[the remark this answers is not in the document]
Comment (negz) [reply docs/adding-a-managed-service.md]: Yeah, I don't think the glossary is merged yet. Will replace the placeholders before this leaves draft.

File: docs/adding-a-managed-service.md [hunk 5]
```diff
@@ -0,0 +1,508 @@
...
+Crossplane does not require controllers to be written in any particular
+language. The Kubernetes API server is our API boundary, so any process capable
+of [watching the API server] and updating resources can be a Crossplane
+controller.
+
+## Getting Started
+
+At the time of writing all Crossplane Services controllers are written in Go,
+and built using [kubebuilder] v0.2.x and [crossplane-runtime].  Per [What Makes
+a Crossplane Managed Service] it is possible to write a controller using the any
```
[the remark this answers is not in the document]
Comment (negz) [reply docs/adding-a-managed-service.md]: I hoped AnyLang was a thing so I could make a snarky retort, but it is not. :(

File: docs/adding-a-managed-service.md [hunk 6]
```diff
@@ -0,0 +1,568 @@
+---
+title: Adding a Managed Service
+toc: true
+weight: 460
```
Comment (prasek) [inline docs/adding-a-managed-service.md]: weight: 720
indent: true

so it's indented under the updated side-nav in #816 with "Developer Guide" at 710: https://github.com/crossplaneio/crossplane/blob/7ecc2df467b5c0683fe257a7cc1d316b76a9445a/docs/developer-guide.md

File: docs/services-developer-guide.md [hunk 7]
```diff
@@ -0,0 +1,1044 @@
...
+possibly even a completely new infrastructure stack. Please do not hesitate to
+[reach out] to the Crossplane maintainers and community for help designing and
+implementing support for new managed services. [#sig-services] would value any
+feedback you may have about the services development process!
+
+[What Makes a Crossplane Managed Service]: #what-makes-a-crossplane-managed-service
+[managed resource]: concepts.md#managed-resource
+[resource claim]: concepts.md#resource-claim
+[resource class]: concepts.md#resource-class
+[dynamic provisioning]: concepts.md#dynamic-and-static-provisioning
```
Comment (negz) [inline docs/services-developer-guide.md]: These concepts links are speculative, because that document does not yet exist.

**Pull request merged**